### PR TITLE
feat(Core/Crypto): add support `OpenSSL 3.0`

### LINKED ISSUE #13159: Feature: openssl 3
### Describe your feature request or suggestion in detail

Please somebody finally change openssl to be version 3. Version 3 is out for over a year it is somewhat common issue among people installing AC especially on ubuntu 22.04.

There was this attempt https://github.com/azerothcore/azerothcore-wotlk/pull/11600 but it has been closed months ago for some reason. There is also reference to other repo seems to me like it is pretty much done it just needs to be modified for AC.

This should also include PR to be made to wiki changing openssl references to reflect that now AC is on v3

### Describe a possible solution to your feature or suggestion in detail

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Winfidonarleyan**
## Changes Proposed:
- Add support library `OpenSSL >= 3.0`

## Issues Addressed:
- Closes #13159

## SOURCE:
- Of course TrinityCore and then my core https://github.com/WarheadCore/WarheadBand/commit/d1172fd35f28e5a0731b6e8c0d577ba3f008324c

## Tests Performed:
- Tested with my core in `Ubuntu 22.04` in WSL mode (OpenSSL 3.0)
- Tested with `OpenSSL 1.1` in linux
image
- Tested in `Ubuntu 22.04` by @alentoghostflame
- Tested in `Win10` and `Win11` by me and @heyitsbench

## How to Test the Changes:
1. Set you lib `OpenSSL` to version `3.0`
2. Just try enter world

## TODO:
- [x] https://github.com/azerothcore/wiki/pull/818
- [x] https://github.com/azerothcore/wiki/pull/819
- [x] https://github.com/azerothcore/wiki/pull/820
- [x] https://gist.github.com/MellianStudios/84471e0fd639964415d11fdc039cb3b0

## How to Test AzerothCore PRs

When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by Winfidonarleyan** [comment]
@heyitsbench can test this?

**Comment by heyitsbench** [comment]
Tested on Windows 10 x64 with a rebuild between tests. [Support for OpenSSL 1.1.1[X] is maintained](https://user-images.githubusercontent.com/38561765/194760507-6ff8fcc3-a64b-42c4-a417-72bac9789ffd.png). [OpenSSL 3.0.5 had an assertion fail upon logging into the server on the authserver.](https://user-images.githubusercontent.com/38561765/194760976-56c8e74d-c284-4562-834e-dac9dd599e40.png)

[Crashes.zip](https://github.com/azerothcore/azerothcore-wotlk/files/9741739/Crashes.zip)

**Comment by Winfidonarleyan** [comment]
> Tested on Windows 10 x64 with a rebuild between tests. [Support for OpenSSL 1.1.1[X] is maintained](https://user-images.githubusercontent.com/38561765/194760507-6ff8fcc3-a64b-42c4-a417-72bac9789ffd.png). [OpenSSL 3.0.5 had an assertion fail upon logging into the server on the worldserver](https://user-images.githubusercontent.com/38561765/194760976-56c8e74d-c284-4562-834e-dac9dd599e40.png).
> 
> [4cebaa13fb6e_worldserver.exe_[9-10_9-57-46]](https://gist.github.com/heyitsbench/bee7d5a3ec43967d0c4373c95a322736) [Crashes.zip](https://github.com/azerothcore/azerothcore-wotlk/files/9741739/Crashes.zip)

can test in linux?

**Comment by Winfidonarleyan** [comment]
@heyitsbench i found fix.
- Copy this lib to `worldserver.exe` dir

image

```
WarheadCore rev. d85122cc1b49+ 2022-10-09 18:45:42 +0300 (master branch) (Win64, RelWithDebInfo, Static)
Using SSL version: OpenSSL 3.0.5 5 Jul 2022 (library: OpenSSL 3.0.5 5 Jul 2022)
Using Boost version: 1.80.0
Using MySQL version: 101001
Using CMake version: 3.24.2
Compiled on: Windows
Worldserver listening connections on port %hu
Realmlist (Realm Id: 1) configured in port %hu
VMAPs status: Enabled. LineOfSight: true, getHeight: true, indoorCheck: true
MMAPs status: Enabled
maps directory located in F:/Dump/Core/data_15/maps. Total size: 291014951 bytes
vmaps directory located in F:/Dump/Core/data_15/vmaps. Total size: 658130721 bytes
mmaps directory located in F:/Dump/Core/data_15/mmaps. Total size: 2320900676 bytes
Using ruRU DBC Locale as default. All available DBC locales: ruRU
Using World DB: ACDB 335.8-dev
AuthDatabase queue size: 0
CharacterDatabase queue size: 0
WorldDatabase queue size: 0
```

**Comment by heyitsbench** [comment]
> @heyitsbench i found fix.
> 
> * Copy this lib to `worldserver.exe` dir

Still in the middle of building with WSL (taking longer than I thought 😛), but [copying that `legacy.dll` seemed to do the trick](https://user-images.githubusercontent.com/38561765/194770055-c724043a-13af-469f-a5d8-22ff2ef5d62c.png)!

**Comment by Winfidonarleyan** [comment]
> > @heyitsbench i found fix.
> > 
> > * Copy this lib to `worldserver.exe` dir
> 
> Still in the middle of building with WSL (taking longer than I thought 😛), but [copying that `legacy.dll` seemed to do the trick](https://user-images.githubusercontent.com/38561765/194770055-c724043a-13af-469f-a5d8-22ff2ef5d62c.png)!

Good, thanks. I will make warnings for windows

**Comment by heyitsbench** [comment]
> Good, thanks. I will make warnings for windows

Can probably just change the wiki guide to include one more file to copy, since Windows already has to copy a few other files from OpenSSL.

**Comment by Winfidonarleyan** [comment]
> > Good, thanks. I will make warnings for windows
> 
> Can probably just change the wiki guide to include one more file to copy, since Windows already has to copy a few other files from OpenSSL.

And this :D

**Comment by alentoghostflame** [comment]
image
Tested on an Ubuntu 22.04 server!

**Comment by Winfidonarleyan** [comment]
> image Tested on an Ubuntu 22.04 server!

Thanks for testing

**Comment by Winfidonarleyan** [comment]
@heyitsbench can retest in win with new assert messages?

**Comment by heyitsbench** [comment]
Getting compile errors.
image

**Comment by Winfidonarleyan** [comment]
fixed

**Comment by heyitsbench** [comment]
With no `legacy.dll` in the build directory:
[authserver](https://user-images.githubusercontent.com/38561765/194773425-5ba5ffe3-11ff-4bdf-bdc3-7020c77bdb1b.png)
[worldserver](https://user-images.githubusercontent.com/38561765/194773433-a6d4ceff-1f52-4748-9e79-96109b5424d4.png)

With `legacy.dll` in the build directory:
image

**Comment by Winfidonarleyan** [comment]
> With no `legacy.dll` in the build directory: [authserver](https://user-images.githubusercontent.com/38561765/194773425-5ba5ffe3-11ff-4bdf-bdc3-7020c77bdb1b.png) [worldserver](https://user-images.githubusercontent.com/38561765/194773433-a6d4ceff-1f52-4748-9e79-96109b5424d4.png)
> 
> With `legacy.dll` in the build directory: image

Very good and big thanks for testing

**Comment by heyitsbench** [comment]
Worth pointing out, it seems that all that's required to copy to the build directory from OpenSSL *is* that legacy.dll, the libcrypto-x.dll and libssl-x.dll don't seem to be required. Don't know if that's the case for OpenSSL 1.1.1x with this PR as well, but it is for 3.0.5.

File: src/common/Cryptography/CryptoRandom.cpp [hunk 1]
```diff
@@ -22,5 +22,5 @@
 void Acore::Crypto::GetRandomBytes(uint8* buf, size_t len)
 {
     int result = RAND_bytes(buf, len);
-    ASSERT(result == 1);
+    ASSERT(result == 1, "Not enough randomness in OpenSSL's entropy pool. What in the world are you running on?");
```
Comment (Kitzunu) [inline src/common/Cryptography/CryptoRandom.cpp:25]: kek

Comment (Winfidonarleyan) [reply src/common/Cryptography/CryptoRandom.cpp:25]: from TC :D

**Review by Kitzunu — APPROVED** [review]

**Comment by Winfidonarleyan** [comment]
> Worth pointing out, it seems that all that's required to copy to the build directory from OpenSSL _is_ that legacy.dll, the libcrypto-x.dll and libssl-x.dll don't seem to be required. Don't know if that's the case for OpenSSL 1.1.1x with this PR as well, but it is for 3.0.5.

Interesting

**Comment by Winfidonarleyan** [comment]
I think need stress test with players (1day +)

**Comment by Helias** [comment]
Interesting PR, good job :rocket:

File: src/common/Cryptography/BigNumber.cpp [hunk 2]
```diff
@@ -1,11 +1,25 @@
 /*
- * Copyright (C) 2016+  AzerothCore <www.azerothcore.org>, released under GNU GPL v2 license, you may redistribute it and/or modify it under version 2 of the License, or (at your option), any later version.
- * Copyright (C) 2021+  WarheadCore <https://github.com/WarheadCore>
+ * This file is part of the AzerothCore Project. See AUTHORS file for Copyright information
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Affero General Public License as published by the
```
Comment (heyitsbench) [inline src/common/Cryptography/BigNumber.cpp:5]: Relicensing.

File: src/common/Cryptography/BigNumber.h [hunk 3]
```diff
@@ -1,14 +1,25 @@
 /*
- * Copyright (C) 2016+  AzerothCore <www.azerothcore.org>, released under GNU GPL v2 license, you may redistribute it and/or modify it under version 2 of the License, or (at your option), any later version.
- * Copyright (C) 2021+  WarheadCore <https://github.com/WarheadCore>
+ * This file is part of the AzerothCore Project. See AUTHORS file for Copyright information
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Affero General Public License as published by the
```
Comment (heyitsbench) [inline src/common/Cryptography/BigNumber.h:5]: Relicensing.

File: src/common/Cryptography/BigNumber.cpp (hunk 2 above)
Comment (Winfidonarleyan) [reply src/common/Cryptography/BigNumber.cpp:5]: All change files, yes

**Comment by acidmanifesto** [comment]
Labeling as WIP so someone does not accidentially hit the merge button

**Comment by alfieo** [comment]
This works for me and completed my install without any error

**Comment by MikaMauger** [comment]
Same here, it works nicely. There are just some warning about MD5 deprecation as stated [here](https://github.com/azerothcore/azerothcore-wotlk/issues/13159#issuecomment-1287902747).

**Comment by IntelligentQuantum** [comment]
@acidmanifesto with 20 player works fine

**Comment by heyitsbench** [comment]
If this is ready for and needs another test, [here's a successful CMake/build log](https://gist.github.com/heyitsbench/1b390958118defa349f1533a9b62cc57), and here's a [`.server debug`](https://user-images.githubusercontent.com/38561765/203462354-d086f20c-6195-481e-82ba-87aab4378f04.png).

**Comment by Winfidonarleyan** [comment]
> If this is ready for and needs another test, [here's a successful CMake/build log](https://gist.github.com/heyitsbench/1b390958118defa349f1533a9b62cc57), and here's a [`.server debug`](https://user-images.githubusercontent.com/38561765/203462354-d086f20c-6195-481e-82ba-87aab4378f04.png).

thanks for test again

**Comment by FrancescoBorzi** [comment]
good job @Winfidonarleyan and thanks  @alentoghostflame and @heyitsbench for testing this!

**Pull request merged**